pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go

```diff
@@ -0,0 +1,272 @@
+// - caches are avoided at all times, we read straight from storage;
+// - MayResolve* methods return zero values if nothing could be found;
+// - MayResolve* methods ignore dropped or offline descriptors;
+// - MustReadDescriptor does not;
+// - MustReadDescriptor panics if the descriptor was not found.
+type CatalogReader interface {
+	tree.TypeReferenceResolver
+	tree.QualifiedNameResolver
+
+	// MayResolveDatabase looks up a database by name.
+	MayResolveDatabase(ctx context.Context, name tree.Name) catalog.DatabaseDescriptor
+
+	// MayResolveSchema looks up a schema by name.
+	MayResolveSchema(ctx context.Context, name tree.ObjectNamePrefix) (catalog.DatabaseDescriptor, catalog.SchemaDescriptor)
+
+	// MayResolveTable looks up a table by name.
+	MayResolveTable(ctx context.Context, name tree.UnresolvedObjectName) (catalog.ResolvedObjectPrefix, catalog.TableDescriptor)
+
+	// MayResolveType looks up a type by name.
+	MayResolveType(ctx context.Context, name tree.UnresolvedObjectName) (catalog.ResolvedObjectPrefix, catalog.TypeDescriptor)
+
+	// ReadObjectNamesAndIDs looks up the namespace entries for a schema.
+	ReadObjectNamesAndIDs(ctx context.Context, db catalog.DatabaseDescriptor, schema catalog.SchemaDescriptor) (tree.TableNames, descpb.IDs)
+
+	// MustReadDescriptor looks up a descriptor by ID.
+	MustReadDescriptor(ctx context.Context, id descpb.ID) catalog.Descriptor
+}
+
+// AuthorizationAccessor for checking authorization (e.g. desc privileges).
+type AuthorizationAccessor interface {
+
+	// CheckPrivilege verifies that the current user has `privilege` on
+	// `descriptor`.
+	CheckPrivilege(
+		ctx context.Context, descriptor catalog.Descriptor, privilege privilege.Kind,
+	) error
+
+	// HasAdminRole verifies if a user has an admin role.
+	HasAdminRole(ctx context.Context) (bool, error)
+
+	// HasOwnership returns true iff the role, or any role the role is a member
+	// of, has ownership privilege of the desc.
+	HasOwnership(ctx context.Context, descriptor catalog.Descriptor) (bool, error)
+}
+
+// BuilderState encapsulates the state of the planned schema changes, hiding
+// its internal state to anything that ends up using it and only allowing
+// state changes via the provided methods.
+type BuilderState interface {
+
+	// AddNode adds a node into the NodeAccumulator.
+	AddNode(dir scpb.Target_Direction, elem scpb.Element)
+
+	// ForEachNode iterates over the accumulated notes in the NodeAccumulator.
+	ForEachNode(fn func(status scpb.Status, dir scpb.Target_Direction, elem scpb.Element))
+
+	// IncrementSubWorkID increments the current subwork ID used for tracking
+	// when a statement does operations on multiple objects or in multiple
+	// commands.
+	IncrementSubWorkID()
+
+	// BuilderStateWithNewSourceElementID returns a BuilderState with an
+	// incremented source element ID, which will be inherited by any element added
+	// using the returned BuilderState.
+	BuilderStateWithNewSourceElementID() BuilderState
```

Comment:
I really don't love that this is exposed. If you need it for the implementation of `buildCtx` expose a private interface in `scbuild` which extends `BuilderState`, or, in lieu of that, just have your one concrete impl of `buildCtx` take a concrete `*scbuild.builderState` and explicitly delegate the two methods here that you actually need to expose to this layer.

Reply:
No, I really want to avoid exposing `builderState` internals. It's dangerous and begging to be abused. Also, since this mutates the state directly, I feel it belongs there. I understand your point that it feels stupid and I'll try to work out a better way of doing the same thing.

Reply:
How about:

```go
type internalBuilderState interface {
     scbuildstmt.BuilderState
     
     forEachNode(/* ... */)
     withNewSourceElementID() internalBuilderState
}

type buildCtx struct {
	scbuildstmts.Dependencies
	internalBuilderState
}
```

Reply:
We could do something like that but I think I came up with something nicer to replace `BuilderStateWithNewSourceElementID` which I'll push right now.